### DISCUSSION TITLE
Switching to swanctl.conf and confused about proposals

### DISCUSSION THREAD

**Discussion by sangdrax8** [strongswan/strongswan#460] [Q&A] [answered]
I am looking at the proposals section of my converted config, and it seems I can no longer use ! to lock the acceptable proposals to only my specified list?  So I have tried adding only one proposal, yet I seem to be connecting with something not from the list.  Specifically I am looking at the phase 2 since that is where the PFS is giving me issues during re-key.

My configuration is as follows:
```
connections {
  
   test {
      local_addrs  = %defaultroute
      remote_addrs = 10.10.10.48
      if_id_in = 02
      if_id_out = 02
      proposals = aes256-sha256-modp8192
      local {
         auth = eap-tls
         certs = cert.pem
      }
      remote {
        id = 10.10.10.48
        auth = pubkey
      }
      children {
         testnet {
            local_ts  = 0.0.0.0/0
            remote_ts = 0.0.0.0/0
            esp_proposals = aes256-sha256-modp8192
            set_mark_out = 220
            updown = /etc/swanctl/iprule2.sh iptables
         }
      }
   }
}
```

When I connect, it seems the phase1 is acceptable, but the phase 2 doesn't negotiate DH groups.  I then notice issues during re-key as it seems the server expects PFS and we fail to connect.  In my ipsec.conf config I used the exact same settings but locked them, and it would negotiate DH groups after I did that.

```
      esp=aes256-sha256-modp8192!
```

When I connect to my server with the above swanctl.conf, it does connect but phase 2 is not what I have listed.
```
  AES_CBC-256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_8192
  established 2s ago, rekeying in 14098s
  testnet: #277, reqid 16, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_256_128
    installed 2s ago, rekeying in 3355s, expires in 3958s
    in  c47b3c6d (-|0x00000002),      0 bytes,     0 packets
    out c61b38cc (-|0x00000002),  10634 bytes,    50 packets,     0s ago
    local  10.10.40.18/32
    remote 0.0.0.0/0
```

**Comment by Thermi** [comment, marked as answer]
It's exactly what you configured. The PFS DH group in IKEv2 is only transmitted when rekeying. It's the exact same between swanctl and "ipsec".

**Reply by sangdrax8** [reply]
wow, thank you your right.  I can just issue swanctl -R --child testnet and after the re-key the algorithm now shows up....

```
  testnet: #283, reqid 21, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_256_128/MODP_8192
    installed 2s ago, rekeying in 3462s, expires in 3958s
    in  cce06bfe (-|0x00000002),      0 bytes,     0 packets
    out ca00fafd (-|0x00000002),      0 bytes,     0 packets
```

I guess I'll have to see if I have additional issues, but this seems to be working.

**Reply by sangdrax8** [reply]
Hm, so that worked on a test case.  I am now testing with another server (pfsense) and when I set the phase 2 the same way, I do get connected again, but when I re-key I still don't negotate DH groups.  It does say rekey was successful.

```
root@brians-XPS-13-9310:/etc/swanctl/conf.d# swanctl -R --child net
rekey completed successfully
root@brians-XPS-13-9310:/etc/swanctl/conf.d# swanctl  -l
test: #1, ESTABLISHED, IKEv2, f37a13d87a2d9dc6_i* 64a3a7fa7c0fb687_r
  local  'test' @ 10.10.20.40[4500] [10.10.40.18]
  remote '10.10.10.10' @ 10.10.10.10[4500]
  AES_CBC-256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_8192
  established 94s ago, rekeying in 12928s
  net: #1, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_256_128
    installed 94s ago, rekeying in 3405s, expires in 3866s
    in  cba6c3ce (-|0x00000002), 9347653 bytes,  6781 packets,     1s ago
    out c12aef7c (-|0x00000002), 128623 bytes,   972 packets,     1s ago
```

I guess I would expect either the re-key to use DH, or fail.  It seems that forcing the re-key it still connects even when not matching the settings?

**Reply by sangdrax8** [reply]
ah, you know what, the phase 2 doesn't say it was installed a few seconds ago, it STILL says it was 94s ago the same as the phase one.  So perhaps th econfusion here is the "rekey completed successfully" part?  Did it actually NOT re-key because it didn't match and just kept the initial configuration which matches but doesn't have DH parameters yet.

**Reply by sangdrax8** [reply]
Alright, I can confirm that if I remove the DH groups from my config, and do a re-key (against my pfsense test) it still says re-key successful, however this time around it also shows the installed time resets so it actually did rekey.

I am going to guess this is just a bug of sorts in with swanctl -R --child XXX saying success when it was in fact not successful.  I can look at my server since I got my other strongswan server to work.  I probably need to evaluate the proposals there to include DH.

**Reply by Thermi** [reply]
Configure logging as shown in #196 and find out!

**Reply by tobiasbrunner** [reply]
As @Thermi suggested, you should read the logs to find out what actually happened. But yes, the `--rekey` command does not actually wait for any rekeyings to complete. All it does is queue rekey jobs that initiate rekeyings asynchronously (could be multiple if you do this by name and have several IKE/CHILD_SAs). In that regard, the output by `swanctl` is probably confusing. If anything, it refers to the `rekey()` VICI command, not any actual rekeyings.
